nvflare/app_common/streamers/object_retriever.py

```diff
@@ -0,0 +1,284 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import threading
+import time
+import uuid
+from abc import ABC, abstractmethod
+from typing import Any
+
+from nvflare.apis.event_type import EventType
+from nvflare.apis.fl_component import FLComponent
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import ReturnCode, Shareable, make_reply
+from nvflare.apis.streaming import StreamContext
+from nvflare.security.logging import secure_format_exception
+
+RETRIEVER_TX_ID = "_rtr_tx_id_"
+
+_SHORT_WAIT = 0.1
+
+
+class _Waiter(threading.Event):
+    def __init__(self):
+        super().__init__()
+        self.result = None
+
+    def set_result(self, rc: str, data: Any):
+        self.result = (rc, data)
+
+
+class ObjectRetriever(FLComponent, ABC):
+    """This is the base class for object retrieval with streaming. The retrieval works as follows:
+    - The requesting site initiates the process by sending a data request to the site that has the data;
+    - The requesting site then waits for the data to be completely received;
+    - Once the data request is received, the data owner site streams the data to the requesting site;
+    - During the streaming process, the requesting site keeps checking for the completion of the streaming until
+    either the data is completely received, or timed out, or aborted.
+    """
+
+    def __init__(
+        self,
+        topic: str = None,
+    ):
+        FLComponent.__init__(self)
+        class_name = self.__class__.__name__
+        if not topic:
+            topic = class_name
+        self.topic = topic
+        self.stream_channel = class_name
+        self.tx_table = {}
+
+    @abstractmethod
+    def register_stream_processing(
+        self,
+        channel: str,
+        topic: str,
+        fl_ctx: FLContext,
+        stream_done_cb,
+        **cb_kwargs,
+    ):
+        """Object requester side, which will receive data stream.
+        This is called to register the status_cb for received stream.
+
+        Args:
+            channel: stream channel
+            topic: stream topic
+            fl_ctx: FLContext object
+            stream_done_cb: the stream_done callback to be registered
```

Comment:
The signature of stream_done_cb should be described here

Reply:
It's the same stream_done_cb as documented in apis.streaming.